common/src/main/java/org/apache/apex/common/util/JarHelper.java

```diff
@@ -0,0 +1,190 @@
+import java.util.jar.Manifest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.commons.compress.archivers.ArchiveException;
+import org.apache.commons.compress.archivers.ArchiveOutputStream;
+import org.apache.commons.compress.archivers.ArchiveStreamFactory;
+import org.apache.commons.compress.archivers.jar.JarArchiveEntry;
+import org.apache.commons.io.IOUtils;
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+
+@InterfaceAudience.Public
+@InterfaceStability.Evolving
+public class JarHelper
+{
+  private static final Logger logger = LoggerFactory.getLogger(JarHelper.class);
+
+  private final Map<URL, String> sourceToJar = new HashMap<>();
+
+  public static String createJar(String prefix, File dir, boolean deleteOnExit) throws IOException
+  {
+    if (!dir.exists() || !dir.isDirectory()) {
+      throw new IllegalArgumentException(String.format("dir %s must be an existing directory.", dir));
+    }
+    File temp = File.createTempFile(prefix, ".jar");
+    if (deleteOnExit) {
+      temp.deleteOnExit();
+    }
+    new JarCreator(temp).createJar(dir);
+    return temp.getAbsolutePath();
+  }
+
+  public String getJar(Class<?> jarClass)
+  {
+    String jar = null;
+    final CodeSource codeSource = jarClass.getProtectionDomain().getCodeSource();
+    if (codeSource != null) {
+      URL location = codeSource.getLocation();
+      jar = sourceToJar.get(location);
+      if (jar == null) {
+        // don't create jar file from folders multiple times
+        if ("jar".equals(location.getProtocol())) {
+          try {
+            location = ((JarURLConnection)location.openConnection()).getJarFileURL();
+          } catch (IOException e) {
+            throw new AssertionError("Cannot resolve jar file for " + jarClass, e);
+          }
+        }
+        if ("file".equals(location.getProtocol())) {
+          jar = location.getFile();
+          final File dir = new File(jar);
+          if (dir.isDirectory()) {
+            try {
+              jar = createJar("apex-", dir, false);
+            } catch (IOException e) {
+              throw new AssertionError("Cannot resolve jar file for " + jarClass + ". URL " + location, e);
+            }
+          }
+        } else {
+          throw new AssertionError("Cannot resolve jar file for " + jarClass + ". URL " + location);
+        }
+        sourceToJar.put(location, jar);
+        logger.debug("added sourceLocation {} as {}", location, jar);
+      }
+      if (jar == null) {
+        throw new AssertionError("Cannot resolve jar file for " + jarClass);
+      }
+    }
+    return jar;
+  }
+
+  private static class JarCreator
+  {
+
+    private final ArchiveOutputStream aos;
+
+    private JarCreator(File file) throws IOException
+    {
+      try {
+        aos = new ArchiveStreamFactory().createArchiveOutputStream(ArchiveStreamFactory.JAR, new FileOutputStream(file));
```

Comment:
No major concern. I just thought that when adding this new we can go with the JDK provided solution instead of relying on the older code that was necessary before nio.